### PR TITLE
Features/data structure for nurbs learning

### LINKED ISSUE #136: Improve NURBS Surface Representation for Heliostats with Custom Optimization Directions
### Description

In the current implementation of heliostat surfaces using NURBS in the differentiable raytracer, there are two major limitations:

1. **Restricted Optimization Direction (z-only) and same Optimization Direction for all Control Points**  
   At present, all NURBS control points are optimized only along the z-direction. This is problematic in cases where:
   - The ideal canting of a heliostat facet does not align with the z-axis (e.g., for flat but tilted heliostat facets).
   - The heliostat is designed to focus, and the ideal surface normals vary spatially across control points.
   
2. **Unified Representation of Ideal Surface and Surface Errors**  
   It is currently difficult to store both the ideally canted heliostat surface and its corresponding surface errors meaningfull within a single instance.

---

### Proposed Solution

Introduce a more flexible surface fitting scheme by allowing per-control-point optimization directions. Specifically:

- Define the **ideal heliostat surface** using:
  - `c_ideal`: the ideal 3D coordinates of each control point.
  - `n`: an (optional) normal vector for each control point that defines the local optimization/fitting direction.
  
- During surface fitting, optimize a scalar deviation `h` per control point along its corresponding normal `n`.  
  The actual (error-affected) control point is then computed as: `c_real = c_ideal + h * n`

This formulation ensures that surface errors are applied along meaningful, physically motivated directions—typically the local ideal surface normal—while maintaining a clean separation between the ideal surface and the error model.

Image
---

### Potential Problems

- Ideal control points must be defined. If they are unknown, fallback strategies can be used—e.g., assuming a flat base surface (z = 0, normals along the z-direction), or fitting an ideal NURBS surface (degree 1 for flat heliostats, degree 2 for focusing heliostats) to the measured point cloud.

---

### Benefits

- Supports general canting geometries and focusing shapes.
- Surface errors are represented more accurately along normal directions instead of a fixed z-direction.

### PR DESCRIPTION
**Pull request by MarleneBusch**
# NURBS surface learning from measured flux density distributions

I have implemented the feature to learn heliostat surfaces from bitmaps.
I have adapted the NURBS to handle multiple surfaces at once -> batched tensors.
I have adapted the scenario.h5 files so that it saves variables like ``nurbs_degree_e`` and ``nurbs_degree_n`` together, ``aim_points`` and ``evaluation_points`` were removed.
The ``surface_converter`` is now called ``surface_generator`` and is able to also generate ideal nurbs control points for the .h5 file, while also still offering the possibility to fir the control points to deflectometry.
I added a ``surface_reconstructor`` that handles the nurbs surface reconstruction from measured fluxes. There is no regularisation for the optimizer implemented yet, the optimizer optimizes in all 3 directions and canting is done once for the initial ideal control points but not within the optimization loop, there is also no decanting.
I had to remove all torch.nn.Parameter, the parameter selection is still handled automatically, depending on wether the kinematic calibration or the surface learning is called.
I changed the structure of the ``kinematic_optimizer`` in a way that the setup happens a bit more automatically, the same ideas were applied to the ``surface_reconstructor``.
I added tests and a tutorial.py.
The read the docs tutorials have not been updated on purpose, as I want to do that in a seperate branch, before the release but after all code is in the main, I hope this is okay.
 

Fixes #136 #70 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by MarleneBusch** [comment]
This PR will also now Fixes #65
I have added epsilon as a parameter to the nurbs evaluation function and found out it can go as low as 1e-7 which i have set as the default value.

**Review by kalebphipps — APPROVED** [review]
Thanks for all the changes!

I have made a few comments but I am happy for this to be merged without an additional review once the changes are addressed.

The main points I think we need to address:
- With the current parallel structure it is not always clear what shape the tensors have, especially when they are expanded, squeezed etc. Therefore it would be really helpful to document these shapes throughout the code (in the docstrings and then my major reshapes as comments in the code). This would make it easier for me to verify that things are working as expected. It would also make it easier for additional features in ARTIST if we always know what shapes are expected and can conform to them.
- I am a bit worried about the amount of data stored in the Git at the moment. As far as I can tell we have a git size now closing in on 150-200mb and if this keeps expanding it will eventually make everything slower and a pain to deal with. I know we need data for tests and for certain tutorials, but if data is not truly required and it is possible to provide a script to download this data/generate that scenario, then I would prefer this.

Other than that everything is looking good - thanks for all the work!

File: tutorials/data/scenarios/test_scenario_paint_multiple_heliostat_groups_ideal_500_cp.h5
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply tutorials/data/scenarios/test_scenario_paint_multiple_heliostat_groups_ideal_500_cp.h5:1]: i have removed it.

File: artist/scenario/surface_generator.py [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+        return surface_config
+
+    def generate_ideal_surface_config(
+        self,
+        facet_translation_vectors: torch.Tensor,
+        canting: torch.Tensor,
+        device: torch.device | None = None,
+    ) -> SurfaceConfig:
+        """
+        Generate an ideal surface configuration.
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/scenario/surface_generator.py:372]: done.

File: artist/scenario/surface_generator.py [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+
+    def perform_canting_and_translation(
+        self,
+        points: torch.Tensor,
+        translation: torch.Tensor,
+        canting: torch.Tensor,
+        device: torch.device | None = None,
+    ) -> torch.Tensor:
+        """
+        Perform the canting rotation and facet translation.
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/scenario/surface_generator.py:457]: done.

File: artist/util/nurbs.py [hunk 3]
```diff
@@ -76,19 +68,25 @@ def __init__(
...
-        self.degree_n = degree_n
-        self.evaluation_points_e = evaluation_points_e
-        self.evaluation_points_n = evaluation_points_n
-        self.control_points = torch.nn.Parameter(control_points)
-        self.knot_vector_e, self.knot_vector_n = self.calculate_knots(device=device)
+        self.degrees = degrees
+        self.control_points = control_points
+        self.uniform = uniform
+        self.number_of_surfaces = self.control_points.shape[0]
+        self.number_of_facets_per_surface = self.control_points.shape[1]
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/util/nurbs.py:75]: I added them, they were indeed missing.

File: artist/scenario/surface_generator.py [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+            control_points=control_points,
+            device=device,
+        )
+
+        # Optimize the control points of the NURBS surface.
+        optimizer = torch.optim.Adam(
+            [nurbs_surface.control_points.requires_grad_()],
+            lr=initial_learning_rate,
+        )
+        scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/scenario/surface_generator.py:186]: I have changed the optimizer and scheduler to be parameters.

File: artist/core/heliostat_ray_tracer.py [hunk 5]
```diff
@@ -245,10 +242,8 @@ def __init__(
             The amount of samples (Heliostats) processed parallel within a single rank (default is 1).
         random_seed : int
             The random seed used for generating the distortions (default is 7).
-        bitmap_resolution_e : int
-            The resolution of the bitmap in the east dimension (default is 256).
-        bitmap_resolution_u : int
-            The resolution of the bitmap in the up dimension (default is 256).
+        bitmap_resolution : torch.Tensor
+            The resolution of the bitmap (default is torch.tensor([256, 256])).
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/core/heliostat_ray_tracer.py:246]: I am not sure if I understand this issue correctly, but i have removed the space.

File: artist/field/surface.py [hunk 6]
```diff
@@ -33,23 +38,30 @@ def __init__(self, surface_config: SurfaceConfig) -> None:
...
-                canting_n=facet_config.canting_n,
+        device = get_device(device=device)
+
+        self.nurbs_facets = []
+
+        for facet_config in surface_config.facet_list:
+            self.nurbs_facets.append(
+                NURBSSurfaces(
+                    degrees=facet_config.degrees,
+                    control_points=facet_config.control_points.unsqueeze(0)
```
[the remark this answers is not in the document]
Comment (MarleneBusch) [reply artist/field/surface.py:54]: done.
